diff --git a/Makefile.am b/Makefile.am
index c9d6729..5d2164f 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -16,9 +16,13 @@ mg_SOURCES = $(OSRCS) $(SRCS) $(OINCS) $(INCS)
 bin_PROGRAMS = mg
 
 docdir = $(datadir)/doc/@PACKAGE@
-doc_DATA = README README.debian README.programmer README.ucsnames \
-	README.auto-execute README.dired README.mg3 README.python-mode README.undo \
-	README.bindings README.fill-paragraph README.reference README.user_macros \
-	README.bsd README.md README.shell-command README.user_modes \
-	README.clike-mode README.misc README.todo
 
+install-data-hook:
+	mkdir -p $(DESTDIR)$(docdir) ; \
+	zip -9rD $(DESTDIR)$(docdir)/READMES.zip \ 
+		README.debian README.programmer README.ucsnames \
+		README.auto-execute README.dired README.mg3 README.python-mode README.undo \
+		README.bindings README.fill-paragraph README.reference README.user_macros \
+		README.bsd README.md README.shell-command README.user_modes \
+		README.clike-mode README.misc README.todo && \
+	zip -9rD $(DESTDIR)$(docdir)/samples.zip bl/*
diff --git a/debian/rules b/debian/rules
index 59ea751..e7bce8d 100755
--- a/debian/rules
+++ b/debian/rules
@@ -23,3 +23,5 @@
 #override_dh_auto_configure:
 #	dh_auto_configure -- \
 #	-DCMAKE_LIBRARY_PATH=$(DEB_HOST_MULTIARCH)
+override_dh_auto_clean:
+	@-make -j1 distclean
